### ISSUE TITLE
FastTree EarlyStoppingRule definition is inconsistent with API

### ISSUE THREAD

**Issue by rogancarr** [dotnet/machinelearning#2520]
In `FastTree`, the `EarlyStoppingRule` is defined as
```cs
[Argument(ArgumentType.Multiple, HelpText = "Early stopping rule. (Validation set (/valid) is required.)", ShortName = "esr", NullName = "<Disable>")]
[TGUI(Label = "Early Stopping Rule", Description = "Early stopping rule. (Validation set (/valid) is required.)")]
public IEarlyStoppingCriterionFactory EarlyStoppingRule;
```

This can be specified like so:
```cs
var fastTreeTrainer = mlContext.Regression.Trainers.FastTree(new 
    Trainers.FastTree.FastTreeRegressionTrainer.Options {
        EarlyStoppingMetrics = 2,
        EarlyStoppingRule = new GLEarlyStoppingCriterion.Arguments()
    });
```

This exposes the `IComponentFactory` way of doing things in the public API, which seems to be inconsistent with the public API. One suggestion is to use an `enum` over the existing options.

That said, this does support custom early stopping methods through implementing  `IEarlyStoppingCriterionFactory`.

**Comment by wschin** [comment]
It will be a part of PR #2753 which will fix #2617.

**Comment by wschin** [comment]
We can't make them `enum` because they all have their own options such as `WindowSize` below.
```csharp
    public sealed class UPEarlyStoppingCriterion : EarlyStoppingCriterion<UPEarlyStoppingCriterion.Options>
    {
        public sealed class Options : OptionsBase, IEarlyStoppingCriterionFactory
        {
            public int WindowSize = 5;
        }
```
